File: julian/rampprofiles.jl
###
#
##= RAMP PROFILES =##
#
###



###   Create a ramp profile structure   ###
#
###
struct RampProfile
    ramp_type::String   # ramp type
    d_max::Float64      # max distance
    t_max::Float64      # max time 
end
###



###
function z_latt(t::Float64, 
                rp::RampProfile)

    t_reduced = t/rp.t_max
    
    if rp.ramp_type == "None"
        z_lattice = 0

    elseif rp.ramp_type == "Linear"
        if t < rp.t_max/2
            z_lattice = 2*rp.d_max * t_reduced^2
        elseif t >= rp.t_max/2
            z_lattice = (-rp.d_max) + (4*rp.d_max * t_reduced) - (2*rp.d_max * t_reduced^2)
        end

    elseif rp.ramp_type == "Minimum Jerk"
        z_lattice = rp.d_max*(10*t_reduced^3 - 15*t_reduced^4 + 6*t_reduced^5)
    end 

    return z_lattice
end 
###



###
function v_latt(t::Float64, 
                rp::RampProfile)
    
    t_reduced = t/rp.t_max

    if rp.ramp_type == "None"
        v_lattice = 0

    elseif rp.ramp_type == "Linear"
        if t < rp.t_max/2
            v_lattice = (4*rp.d_max/rp.t_max^2)*t
        elseif t >= rp.t_max/2
            v_lattice = (4*rp.d_max/rp.t_max^2)*t - (4*rp.d_max/rp.t_max^2)*t
        end
    
    elseif rp.ramp_type == "Minimum Jerk"
        v_lattice = 30*(rp.d_max/rp.t_max)*(t_reduced^2 - 2*t_reduced^3 + t_reduced^4)
    end

    return v_lattice
end
###



###
function a_latt(t::Float64,
                rp::RampProfile)
    
    t_reduced = t/rp.t_max

    if rp.ramp_type == "None"
        a_lattice = 0

    elseif rp.ramp_type == "Linear"
        if t < rp.t_max/2
            a_lattice = 4*rp.d_max/rp.t_max^2
        elseif t >= rp.t_max/2
            a_lattice = -4*rp.d_max/rp.t_max^2
        end
    
    elseif rp.ramp_type == "Minimum Jerk" 
        a_lattice =  60*(rp.d_max/rp.t_max^2)*(t_reduced - 3*t_reduced^2 + 2*t_reduced^3)
    end 

    return a_lattice
end

File: montepython/ramp_profiles.py
import numpy as np
from numba import jit

from constants import *



#=======================================================#
#                       RAMP PROFILES                   #   
#=======================================================# 
#
###


@jit(nopython=True, fastmath=True, cache=True)
def z_latt(ramp_type,
           t, 
           d_max, t_max):

    """
    Computes the derivative of the beam width with respect to z.

    Parameters
    ----------
    t (float): 
        Time. 

    d_max (float): 
        Transport distance.

    t_max (float): 
        Transport time. 

    Returns
    -------
    z_lattice (float): 
        Position of the ODT minimum along the optical axis. 
    """ 
    
    if ramp_type == 0: 
        z_lattice = 0
    
    elif ramp_type == 1: 
        t_reduced = t/t_max
        z_lattice = d_max*(10*(t_reduced)**3 - 15*(t_reduced)**4 + 6*(t_reduced)**5)

    return z_lattice



@jit(nopython=True, fastmath=True, cache=True)
def v_latt(ramp_type,
           t,
           d_max, t_max): 
    
    if ramp_type == 0: 
        v_lattice = 0
        
    elif ramp_type == 1: 
        t_reduced = t/t_max
        v_lattice = 30*(d_max/t_max)*(t_reduced**2 - 2*t_reduced**3 + t_reduced**4)
        
    return v_lattice

###
